### PR TITLE
Performance-optimized Circle Tool

### PR DESCRIPTION
**Pull request by Hiderrr**
The template I used to write the CircleTool.gd class was the RecangleTool.gd class.
Because of this, the code is written in a similar way. I hope it helps (;

**Features:**
  - Well-looking, though not expansive circles (empty, outlines only)
  - UI tool toggle control (created in the same manner as other tools')
  - Number of points on the circle is configurable through CircleTool.gd class
  - While the circle remains in motion (user hasn't released LMB yet), the circle is drawn using a far limited number of points, in order to increase performance (which also gives quite a good looking effect of a "polygonized-circle")
  - Once settled, the final number of points is set and saved

One last thing to do, to make the tool release-ready, is to create a shortcut (C) for the tool

**Future plans:**
  - If I am requested to, I'd more than happily try to implement these features:
    - Changing color of selected strokes to the chosen one
    - UI for the Circle tool, giving additional the user additional control over the number of circle's points in *motion* and *settled* states
    - Making an eraser tool, that actually erases points ((:
    
Regards!
Hiderr

### CODE REVIEW

**Review by mbrlabs — CHANGES_REQUESTED** [review]
Very nice implementation, thanks! I left some comments.
Also in the title you mentioned something about the shift key, but it doen't seem to do anything?

As for the shorcut: `c` is currently used to center the canvas to the mouse position. I think it's best to change that to the `space` key and use `c` for the circle tool (see Project settings > Input map).

**Comment by mbrlabs** [comment]
> Changing color of selected strokes to the chosen one

I think that's already possible, no?

> UI for the Circle tool, giving additional the user additional control over the number of circle's points in motion and settled states

I think it's fine as is. What could be added though is drawing ellipses instead of circles when the shift key is held

> Making an eraser tool, that actually erases points 

That would be great! See this issue for inspiration: https://github.com/mbrlabs/Lorien/issues/74. Basically when the eraser intersects with a bursh it would delete it. If you want to discuss this further let's talk about it in the linked issue

File: lorien/InfiniteCanvas/Tools/CircleTool.gd [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+var sin_arr : Array
+var cos_arr : Array
+
+func _init():
+	sin_arr.resize(360)
+	cos_arr.resize(360)
+	for i in range(0, 360):
+		sin_arr[i] = sin(deg2rad(i))
+		cos_arr[i] = cos(deg2rad(i))
```
[the remark this answers is not in the document]
Comment (Hiderrr) [reply lorien/InfiniteCanvas/Tools/CircleTool.gd:44]: I meant that the tool behaves similarly, as if you hold shift in gimp (:

Thanks for the informative review, I'll commit the requested changes today.

[the remark this answers is not in the document]
Comment (Hiderrr) [reply lorien/InfiniteCanvas/Tools/CircleTool.gd:44]: Also a note here: for loop cannot be simplified, as it uses the step passed to the function for limiting the number of points on the circle.

[the remark this answers is not in the document]
Comment (mbrlabs) [reply lorien/InfiniteCanvas/Tools/CircleTool.gd:44]: I just meant the loop in the constructor (which increments in 1-steps).

**Review by mbrlabs — APPROVED** [review]

**Comment by mbrlabs** [comment]
Code looks good! Also thanks for the ellipse tool. 👍 
I just played around with it a bit and i think it's better to make the ellipse default and enable the circle when holding the shift key. This is because most of the time you want to encircle something and it's easier to do that with an ellipse.

**Comment by Hiderrr** [comment]
> I just played around with it a bit and i think it's better to make the ellipse default and enable the circle when holding the shift key

I do agree, that's actually default behavior in almost all software.

**Comment by Hiderrr** [comment]
I combined the commits to make the tree a little cleaner.
Last commit alters the behavior of SHIFT, so that now it enables "circle mode", and the default mode is "ellipse mode"

**Comment by mbrlabs** [comment]
Great work, thank you very much! :)

**Pull request merged**